rfc/plugins.md

```diff
@@ -0,0 +1,235 @@
+        .then(function(aws) {
+          return {
+            log: aws.log,
+            bucket: aws.bucket
+          };
+        });
+    }
+  }
+}
+```
+
+If a promise from any of the plugins is rejected then the deployment
+pipeline will stop and ember-cli-deploy will exit. Returned rejected
+promises are treated as unrecoverable errors.
+
+### Accessing data
+
+A plugin can access data from another plugin in two ways, by name or by
+hook.
+
+```javascript
+S3Plugin = {
+  hooks: {
+    assets: function() {
+      return this.uploadToS3()
+        .then(function(aws) {
+          return {
+            log: aws.log
+          }
+        });
+    }
+  }
+}
+```
+
+```javascript
+ChatNotifierPlugin = DeployPlugin.extend({
+  didDeploy: function() {
+    // target a hook
+    var assetLog = this.pluginValue('assets').log;
+
+    // or target a specific plugin
+    // this.pluginValue('s3-plugin:assets');
+
+    this.notifyChat('deploys finished!')
+    this.notifyChat(assetLog);
+  }
+});
+
+```
+
+At this point in the ember-cli-deploy project life cycle I don't think we
+can expect plugin authors to conform to a unified interface. That is to
+say, not all ``asset`` plugins will return the same data. This part of
+the RFC is a huge gray area for me, but I think we need to allow this to
+take place and see how plugin authors adopt it. Basically, give everyone
+an escape hatch and see how they use it before trying to come up with an
+API.
+
+I'm not sure what to name this function, just called it `pluginValue`,
+someone please come up with a better name.
+
+### Commands and tasks
+
+Plugins will have the ability to register their own commands to be used
+outside of the deployment pipeline. These commands will use the
+`Task` object that comes with ember-cli.
+
+```javascript
+S3Plugin = {
+  hooks: { ... },
+
+  commands: {
+    doctor: Task.extend({
+      // checks to make sure S3Plugin is configured correctly
+      run: function() { ... }
+    })
+  }
+}
+```
+
+This task would run as `ember deploy s3-plugin:doctor`.
+
+Is this worth providing or should we just let plugin authors do this
+themselves in their ember-cli-addons?
+
+I think it might be worth exploring because there might be many common
+tasks that these plugins provide. We could have an `ember deploy:doctor`
+that runs the doctor task of each plugin and reports back their return
+values. Things like `dry-run`, `doctor`, or `setup` might be common
+tasks you would find across many different plugins.
+
+### Configuration
+
+Plugin configuration should be kept in `config/deploy.js` and scoped by
+the plugin's name.
+
+```javascript
+module.exports = {
+  production: {
+    "s3-plugin": {
+      bucket: "my-app-production",
+      awsAccessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      awsSecretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
+    }
+  }
+};
+```
+
+This data will then be available to the `S3Plugin` function's as
+`this.pluginConfig`.
+
+
```

Comment:
Can we add a section to this document specifying the hook methods? We can incorporate the results of the conversation from #47